core/src/main/java/com/salesforce/slds/validation/validators/impl/recommendation/MobileFriendlyValidator.java

```diff
@@ -0,0 +1,120 @@
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+package com.salesforce.slds.validation.validators.impl.recommendation;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+import com.salesforce.slds.shared.models.context.Context;
+import com.salesforce.slds.shared.models.core.Bundle;
+import com.salesforce.slds.shared.models.core.Entry;
+import com.salesforce.slds.shared.models.core.HTMLElement;
+import com.salesforce.slds.shared.models.core.Input;
+import com.salesforce.slds.shared.models.recommendation.Action;
+import com.salesforce.slds.shared.models.recommendation.ActionType;
+import com.salesforce.slds.shared.models.recommendation.Item;
+import com.salesforce.slds.shared.models.recommendation.Recommendation;
+import com.salesforce.slds.shared.utils.ResourceUtilities;
+import com.salesforce.slds.validation.validators.interfaces.RecommendationValidator;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Component;
+
+import java.io.InputStream;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+@Component
+@Lazy
+public class MobileFriendlyValidator implements RecommendationValidator, InitializingBean {
+    Map<String, List<String>> componentsRegistry;
+
+    @Override
+    public List<Recommendation> matches(Entry entry, Bundle bundle, Context context) {
+        if (entry.getEntityType() != Entry.EntityType.LWC) {
+            return Lists.newArrayList();
+        }
+
+        return entry.getInputs().stream().filter(input -> input.asElement() != null)
+                .map(Input::asElement)
+                .filter(htmlElement -> !isMobileFriendly(htmlElement))
+                .map(htmlElement -> {
+                    Recommendation.RecommendationBuilder builder = Recommendation.builder()
+                            .input(htmlElement);
+
+                    String tag = htmlElement.getContent().tagName();
+
+                    Action action = Action.builder()
+                            .value(htmlElement.getContent().outerHtml()).range(htmlElement.getRange())
+                            .description(tag
+                                    + " is known to have issues on mobile devices. Consider using a replacement " +
+                                    "or create a custom component to use instead.")
+                            .name(tag)
+                            .actionType(ActionType.NONE)
+                            .build();
+
+                    Item item = new Item(tag, action);
+                    builder.items(Sets.newHashSet(item));
+
+                    return builder.build();
```

Comment:
This feels like another block of code that could be pulled out into a helper function. Indeed, I'm surprised one doesn't exist already? Isn't this basically the response mechanism for any recommendation?

Reply:
Partially, yes. List of recommendations are constructed in `matches` method in the six available validators in the project. Generally they follow the pattern of using a streaming API to use filter, map, and flatmap to finally get list of recommendation. Depending on which validator the combination is slightly different from one another. For this mobile validator as you indicated line 46 and 66 can extracted to a private function of its own.

If you look at `CodeActionConverter.java` I refactored out repeated code into a new method. As we contribute to this project I think we shall find more opportunities to refactor and make the code more compact.